ddtrace/debugging/_encoding.py

```diff
@@ -289,17 +290,82 @@ def _get_fields(obj):
         return {s: _safe_getattr(obj, s) for s in get_slots(obj)}
 
 
-def _captured_value_v2(value, level=1):
-    # type: (Any, int) -> Dict[str, Any]
+def _captured_value_v2(value, level=MAXLEVEL, maxlen=MAXLEN, maxsize=MAXSIZE, maxfields=MAXFIELDS):
+    # type: (Any, int, int, int, int) -> Dict[str, Any]
     _type = type(value)
+
+    if _type in BUILTIN_SIMPLE_TYPES:
+        value_repr = repr(value)
+        return (
+            {
+                "type": _qualname(_type),
+                "value": value_repr,
+            }
+            if len(value_repr) <= maxlen
+            else {
+                "type": _qualname(_type),
+                "value": value_repr[:maxlen],
+                "truncated": True,
+                "size": len(value_repr),
+            }
+        )
+
+    if _type in BUILTIN_CONTAINER_TYPES:
+        if level < 0:
+            return {
+                "type": _qualname(_type),
+                "notCapturedReason": "depth",
+                "size": len(value),
+            }
+
+        if _type is dict:
+            # Mapping
+            data = {
+                "type": "dict",
+                "entries": [
+                    (
+                        _captured_value_v2(k, level=level - 1, maxlen=maxlen, maxsize=maxsize, maxfields=maxfields),
+                        _captured_value_v2(v, level=level - 1, maxlen=maxlen, maxsize=maxsize, maxfields=maxfields),
+                    )
+                    for _, (k, v) in zip(range(maxsize), value.items())
+                ],
+                "size": len(value),
+            }
+
+        else:
+            # Sequence
+            data = {
+                "type": _qualname(_type),
+                "elements": [
+                    _captured_value_v2(v, level=level - 1, maxlen=maxlen, maxsize=maxsize, maxfields=maxfields)
+                    for _, v in zip(range(maxsize), value)
+                ],
+                "size": len(value),
+            }
+
+        if len(value) > maxsize:
+            data["notCapturedReason"] = "collectionSize"
+
+        return data
+
+    # Arbitrary object
+    if level < 0:
+        return {
+            "type": _qualname(_type),
+            "notCapturedReason": "depth",
+        }
+
+    fields = _get_fields(value)
     data = {
         "type": _qualname(_type),
-        "value": _serialize(value, level),
-    }  # type: Dict[str, Any]
-    if _type not in BUILTIN_TYPES:
-        data["fields"] = (
-            {n: _captured_value_v2(v, level - 1) for n, v in _get_fields(value).items()} if level > 0 else None
-        )
+        "fields": {
+            n: _captured_value_v2(v, level=level - 1, maxlen=maxlen, maxsize=maxsize, maxfields=maxfields)
+            for _, (n, v) in zip(range(maxfields), fields.items())
```

Comment:
Is the order of the fields deterministic?

Reply:
No. Should it be? In general we retrieve them from looking into a dictionary, so any form of determinism would require computation 🙁

Reply:
The order we put the fields in the output JSON does not matter but if we limit the number of fields we capture to `maxfields` it would be nice to have snapshots consistent with each other (i.e. get the same x fields all the time, rather than having different fields in different snapshots).

It's not a big deal, if you're concerned about the overhead of sorting the field names alphabetically before deciding which ones we keep we can leave this as is.